### PR TITLE
Feat: Dynamic learning rate decay

### PR DESCRIPTION
**Pull request by maxdinkel**
## Description and Context

This PR adds a step-wise learning rate decay for Stochastic variational inference and the dynamic learning rate decay method from my upcoming paper.

## Related Issues and Merge Requests

* Closes
* Blocks
* Is blocked by
* Follows
* Precedes
* Related to
* Part of
* Composed of

## Checklist

- [ ] My commit messages mention the appropriate issue numbers (advised but optional).
- [ ] I mention the appropriate issue numbers in this MR.
- [ ] I updated documentation where necessary.
- [ ] I updated the README where necessary
- [ ] I have added as (few,) small and fast tests as possible to cover my changes.

## Interested Parties

Possible reviewers:

Other interested parties:

### CODE REVIEW

**Review by gilrrei — APPROVED** [review]
Looks good to me :+1:

File: queens/stochastic_optimizers/learning_rate_decay.py [hunk 1]
```diff
@@ -65,6 +67,120 @@ def __call__(self, learning_rate, params, gradient):
...
+        if self.iteration >= self.decay_interval:
+            learning_rate *= self.decay_factor
+            self.iteration = 0
+            _logger.info("learning_rate=%.2e", learning_rate)
+        else:
+            self.iteration += 1
+        return learning_rate
+
+
+class DynamicLearningRateDecay:
```
Comment (leahaeusel) [inline queens/stochastic_optimizers/learning_rate_decay.py:116]: Is there a specific reason why this class does not inherit from LearningRateDecay like the other classes?

File: queens/stochastic_optimizers/learning_rate_decay.py
Comment (leahaeusel) [inline queens/stochastic_optimizers/learning_rate_decay.py:1]: Would adding some tests for the different learning rate decays make sense? Currently, this file is completely untested.

File: queens/stochastic_optimizers/learning_rate_decay.py (hunk 1 above)
Comment (maxdinkel) [reply queens/stochastic_optimizers/learning_rate_decay.py]: Thanks for the hint!

File: tests/unit_tests/stochastic_optimizers/test_learning_rate_decay.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+    Test invalid values for rho_min.
+    """
+    with pytest.raises(ValueError):
+        DynamicLearningRateDecay(rho_min=rho_min)
+
+
+def test_dynamic_learning_rate_decay():
+    """Test the call method of LogLinearLearningRateDecay class."""
+    np.random.seed(1)
```
Comment (leahaeusel) [inline tests/unit_tests/stochastic_optimizers/test_learning_rate_decay.py:106]: Maybe I'm missing something, but is this really needed for the test?

Comment (maxdinkel) [reply tests/unit_tests/stochastic_optimizers/test_learning_rate_decay.py:106]: I am generating random numbers in line 113, so the random number seed must be set, such that the test is deterministic.

Comment (leahaeusel) [reply tests/unit_tests/stochastic_optimizers/test_learning_rate_decay.py:106]: Oops that's pretty obvious, my bad.

**Review by leahaeusel — APPROVED** [review]

**Review by gilrrei — APPROVED** [review]
